src/planner/binder/expression/bind_lambda.cpp

```diff
@@ -1,38 +1,84 @@
 #include "duckdb/parser/expression/lambda_expression.hpp"
 #include "duckdb/parser/expression/operator_expression.hpp"
 #include "duckdb/planner/expression_binder.hpp"
+#include "duckdb/planner/bind_context.hpp"
+#include "duckdb/parser/expression/columnref_expression.hpp"
+#include "duckdb/planner/binder.hpp"
+#include "duckdb/parser/expression/function_expression.hpp"
 
 namespace duckdb {
 
-// static string ExtractColumnFromLambda(ParsedExpression &expr) {
-//	if (expr.type != ExpressionType::COLUMN_REF) {
-//		throw ParserException("Lambda parameter must be a column name");
-//	}
-//	auto &colref = (ColumnRefExpression &)expr;
-//	if (colref.IsQualified()) {
-//		throw ParserException("Lambda parameter must be an unqualified name (e.g. 'x', not 'a.x')");
-//	}
-//	return colref.column_names[0];
-// }
-
-BindResult ExpressionBinder::BindExpression(LambdaExpression &expr, idx_t depth) {
+BindResult ExpressionBinder::BindExpression(LambdaExpression &expr, idx_t depth, const idx_t lambda_param_count,
+                                            const LogicalType &list_child_type) {
+
 	string error;
 
-	// FIXME: decide from the context whether this is actually a LambdaExpression
-	//  If it is, bind it as a lambda here
-	// // set up the lambda capture
-	// FIXME: need to get the type from the list for binding
-	// lambda_capture = expr.capture_name;
-	// // bind the child
-	// BindChild(expr.expression, depth, error);
-	// // clear the lambda capture
-	// lambda_capture = "";
-	// if (!error.empty()) {
-	// 	return BindResult(error);
-	// }
-	// expr.Print();
-	OperatorExpression arrow_expr(ExpressionType::ARROW, move(expr.lhs), move(expr.rhs));
-	return ExpressionBinder::BindExpression(arrow_expr, depth);
+	if (lambda_param_count != 0) {
+
+		if (expr.lhs.size() != lambda_param_count) {
+			throw BinderException("Invalid number of left-hand side parameters for this lambda function.");
+		}
+		if (!expr.rhs) {
+			throw BinderException("Invalid lambda expression.");
+		}
+
+		// create dummy columns for the lambda parameters (lhs)
+		vector<LogicalType> column_types;
+		vector<string> column_names;
+		vector<string> lhs_strings;
+
+		// positional parameters as column references
+		for (idx_t i = 0; i < expr.lhs.size(); i++) {
+
+			auto column_ref = (ColumnRefExpression &)*expr.lhs[i];
+			if (column_ref.IsQualified()) {
+				throw BinderException("Invalid parameter name '%s': must be unqualified", column_ref.ToString());
+			}
+
+			column_types.emplace_back(list_child_type);
+			column_names.push_back(column_ref.GetColumnName());
+			lhs_strings.push_back(expr.lhs[i]->ToString());
+		}
+
+		// base table alias
+		auto lhs_alias = StringUtil::Join(lhs_strings, ", ");
+		if (lhs_strings.size() > 1) {
+			lhs_alias = "(" + lhs_alias + ")";
+		}
+
+		// create a lambda binding and push it to the lambda bindings vector
+		vector<LambdaBinding> local_bindings;
```

Comment:
I thought about this for a while with Tania, and this approach made the most sense to us, because we need to support nested Lambda's. In order to do so, we need a to be able to access the lambda parameters of (potentially) the lambda expressions that came before this.

Because the expression binder is recursive, we pass on bindings from the parent binder to the child binder. We don't want to duplicate the `vector<LambdaBinding>`, therefore we chose to use a pointer, so we can just set the pointer to the vector in the child binder.

I did something very similar with `MacroBinding` to achieve the same behaviour.

Reply:
This is also why the pointer is reset to `nullptr` once all recursive occurrences of this lambda function are bound.